c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,379 @@
+//!   bb0[2]: _0 = Add(move _2, const 1_i32)
+//!     []: StoreIntoLocal, `a + 1`
+//!     [Rvalue]: Expr, `a + 1`
+//!   bb0[3]: StorageDead(_2)
+//!   bb0[4]: Terminator { source_info: ..., kind: return }
+//! ```
+//!
+//! The statement `_2 = _1` is associated with the expression `a`; the statement
+//! as a whole is storing the result of evaluating `a` into a MIR local, and the
+//! statement's rvalue `_1` represents the expression `a` itself.  Similarly, `_0 =
+//! Add(move _2, const 1)` stores the result of `a + 1` into a local.  If needed,
+//! we could extend the `unlower` pass to also record that `move _2` (a.k.a. `bb0[2]`
+//! `[Rvalue, RvalueOperand(0)]`) is lowered from the `Expr` `a`.
+//!
+//! On `g`, the unlowering map includes the following (among other entries):
+//!
+//! ```text
+//! bb0[5]: Terminator { source_info: ..., kind: _4 = f(move _5) -> [return: bb1, unwind: bb2] }
+//!   []: StoreIntoLocal, `f(y)`
+//!   [Rvalue]: Expr, `f(y)`
+//!   [Rvalue, CallArg(0)]: Expr, `y`
+//! bb1[1]: _0 = Add(move _3, move _4)
+//!   []: StoreIntoLocal, `x + f(y)`
+//!   [Rvalue]: Expr, `x + f(y)`
+//! ```
+//!
+//! The call terminator `_4 = f(move _5)` computes `f(y)` and stores the result
+//! into a local; its rvalue is `f(y)` itself, and the first argument of the rvalue
+//! is `y`.
+
+use crate::panic_detail;
+use crate::rewrite::build_span_index;
+use crate::rewrite::expr::mir_op::SubLoc;
+use crate::rewrite::span_index::SpanIndex;
+use log::*;
+use rustc_hir as hir;
+use rustc_hir::intravisit;
+use rustc_hir::HirId;
+use rustc_middle::hir::nested_filter;
+use rustc_middle::mir::{self, Body, Location};
+use rustc_middle::ty::{DefIdTree, TyCtxt};
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+use std::collections::btree_map::{BTreeMap, Entry};
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+pub struct MirOrigin {
+    pub hir_id: HirId,
+    pub span: Span,
+    pub desc: MirOriginDesc,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+pub enum MirOriginDesc {
+    /// This MIR represents the whole HIR expression.
+    Expr,
+    /// This MIR stores the result of the HIR expression into a MIR local of some kind.
+    StoreIntoLocal,
+}
+
+struct UnlowerVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    mir: &'a Body<'tcx>,
+    span_index: SpanIndex<Location>,
+    /// Maps MIR (sub)locations to the HIR node that produced each one, if known.
+    mir_map: BTreeMap<(Location, Vec<SubLoc>), MirOrigin>,
+}
+
+impl<'a, 'tcx> UnlowerVisitor<'a, 'tcx> {
+    fn location_span(&self, loc: Location) -> Span {
+        self.mir
+            .stmt_at(loc)
+            .either(|stmt| stmt.source_info.span, |term| term.source_info.span)
+    }
+
+    fn record(&mut self, loc: Location, sub_loc: &[SubLoc], ex: &hir::Expr) {
+        self.record_desc(loc, sub_loc, ex, MirOriginDesc::Expr);
+    }
+
+    fn record_desc(
+        &mut self,
+        loc: Location,
+        sub_loc: &[SubLoc],
+        ex: &hir::Expr,
```

Comment:
Is `ex` an abbreviation for `expr` commonly used in rustc?  Otherwise, it's pretty confusing, so I'd prefer `expr`, which we normally use elsewhere.

Reply:
`ex` is the name used in [the HIR `Visitor` trait](https://doc.rust-lang.org/stable/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html#method.visit_expr), and generally seems pretty unambiguous

Reply:
Perhaps it's a MIR vs. HIR distinction?  Fairly odd abbreviation but if it's already used widely it's fine.